lib/compress/zstd_compress.c

```diff
@@ -1055,24 +1055,71 @@ U32 ZSTD_cycleLog(U32 hashLog, ZSTD_strategy strat)
     return hashLog - btScale;
 }
 
+/** ZSTD_dictAndWindowLog() :
+ *  Returns an adjusted window log that is large enough to fit the source and the dictionary.
+ *  The zstd format says that the entire dictionary is valid if one byte of the dictionary
+ *  is within the window. So the hashLog and chainLog should be large enough to reference both
+ *  the dictionary and the window. So we must use this adjusted dictAndWindowLog when downsizing
+ *  the hashLog and windowLog.
+ */
+static U32 ZSTD_dictAndWindowLog(U32 windowLog, U64 srcSize, U64 dictSize)
+{
+    const U64 maxWindowSize = 1ULL << ZSTD_WINDOWLOG_MAX;
+    /* No dictionary ==> No change */
+    if (dictSize == 0) {
+        return windowLog;
+    }
+    assert(srcSize != ZSTD_CONTENTSIZE_UNKNOWN); /* Handled in ZSTD_adjustCParams_internal() */
```

Comment:
maybe this condition worth mentioning in the function description (even though it's just a few lines above)